### DISCUSSION TITLE
ExchangeCodeForSession - Supabase returns HTTP 403 "code challenge does not match..."

### DISCUSSION THREAD

**Discussion by nref** [supabase-community/supabase-csharp#86] [Q&A]
Looking for some help implementing server-side PKCE auth flow in ASP.NET Core.

After calling `Supabase.Client.SignInWithOtp(...)` on the server:

![image](https://github.com/supabase-community/supabase-csharp/assets/2089725/fec8241b-f18f-4756-a330-0eb20b2d9714)

Then looking in the DB `flow_state` table I can see the generated `auth_code` and `code_challenge`:

![image](https://github.com/supabase-community/supabase-csharp/assets/2089725/2c400f70-3b1d-4297-8cac-73e0f09cb0fe)

If I then manually paste `auth_code` and `code_challenge` into my API controller, hash `code_challenge` with `Helpers.GeneratePKCENonceVerifier(...)`, then pass the hash and `auth_code` to `Supabase.Client.ExchangeCodeForSession`, then Supabase returns 

`HTTP 403 "code challenge does not match previously saved code verifier"`

![image](https://github.com/supabase-community/supabase-csharp/assets/2089725/edd793f6-f55a-4221-b94c-a9289bbe6a73)

I would have expected a match.

**Comment by acupofjose** [comment]
The code verifier should be in your variable `result.PKCEVerifier` - which you would use for the session code exchange! (See [here](https://supabase-community.github.io/supabase-csharp/api/Supabase.Gotrue.PasswordlessSignInState.html#Supabase_Gotrue_PasswordlessSignInState_PKCEVerifier))
